keps/sig-auth/3331-structured-config-for-oidc-authentication/README.md

```diff
@@ -0,0 +1,638 @@
+
+This can inform certain test coverage improvements that we want to do before
+extending the production code to implement this enhancement.
+-->
+
+- `<package>`: `<date>` - `<test coverage>`
+
+##### Integration tests
+
+<!--
+This question should be filled when targeting a release.
+For Alpha, describe what tests will be added to ensure proper quality of the enhancement.
+
+For Beta and GA, add links to added tests together with links to k8s-triage for those tests:
+https://storage.googleapis.com/k8s-triage/index.html
+-->
+
+- <test>: <link to test coverage>
+
+##### e2e tests
+
+<!--
+This question should be filled when targeting a release.
+For Alpha, describe what tests will be added to ensure proper quality of the enhancement.
+
+For Beta and GA, add links to added tests together with links to k8s-triage for those tests:
+https://storage.googleapis.com/k8s-triage/index.html
+
+We expect no non-infra related flakes in the last month as a GA graduation criteria.
+-->
+
+- <test>: <link to test coverage>
+
+### Graduation Criteria
+
+#### Alpha
+
+- Feature implemented behind a feature flag
+- Initial e2e tests completed and enabled
+
+#### Beta
+
+- Gather feedback from developers and surveys
+- Complete benchmarks
+
+#### GA
+
+- Add metrics
+- Add a full documentation with examples for the most popular providers, e.g., Okta, Dex, Auth0
+- Migration guide
+- Deprecation warnings for non-stuctured OIDC provider configuration
+
+**Note:** Generally we also wait at least two releases between beta and
+GA/stable, because there's no opportunity for user feedback, or even bug reports,
+in back-to-back releases.
+
+**For non-optional features moving to GA, the graduation criteria must include
+[conformance tests].**
+
+[conformance tests]: https://git.k8s.io/community/contributors/devel/sig-architecture/conformance-tests.md
+
+#### Deprecation
+
+kube-apiserver `--oidc-*` flags require deprecation warnings on the stable release of the feature.
+It is possible to react only to the `--oidc-issuer-url` flag because other flags cannot be enabled separately from this one.
+
+## Production Readiness Review Questionnaire
+
+### Feature Enablement and Rollback
+
+###### How can this feature be enabled / disabled in a live cluster?
+
+- Feature gate
+  - Feature gate name: `StructuredOIDCConfiguration`
+  - Components depending on the feature gate:
+    - kube-apiserver
+
+```go
+FeatureSpec{
+	Default: false,
+	LockToDefault: false,
+	PreRelease: featuregate.Alpha,
+}
+```
+
+###### Does enabling the feature change any default behavior?
+
+No.
+
+###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?
+
+Yes, but disable means also deleting the flag from the kube-apiserver manifest.
+
+###### What happens if we reenable the feature if it was previously rolled back?
+
+No impact.
+
+###### Are there any tests for feature enablement/disablement?
+
+Not required.
```

Comment:
I would imagine some test added as part of https://github.com/kubernetes/kubernetes/issues/110782 would need to enable the `StructuredOIDCConfiguration` feature gate. It probably wouldn't be a stretch for the same test to validate what happens when the gate is disabled.

Reply:
I think even just noting something like this would be sufficient for alpha:
```
Feature enablement/disablement tests will be added as part of https://github.com/kubernetes/kubernetes/issues/110782 
```

Reply:
An example test could be: unit test that demonstrates that when the featuregate is false, the validation function on the Options type reports a failure when the flag is set.

Reply:
Ok, it makes sense. I will add it.